### PR TITLE
Remove internal tables, metadata builder, extension tables only built through static methods

### PR DESCRIPTION
**Pull request by jmaxson-ms**
- Metadata tables are no longer built by default.
- All extension tables (those that require data cookers) must be built through a static build method
- Remove concept of internal tables. There are extension tables built through static build methods, and other tables built by a custom data processor.
- Update tests.

Note that migration documentation will be updated later.

### CODE REVIEW

File: src/Microsoft.Performance.SDK.Runtime/ProcessingSourceExecutor.cs [hunk 1]
```diff
@@ -131,27 +131,6 @@ public void InitializeCustomDataProcessor(ExecutionContext context)
                 return new ExecutionResult(this.Context, this.Processor, e);
             }
 
-            //_CDS_ Should we "delayload" metadata tables?
```
Comment (trgibeau) [inline src/Microsoft.Performance.SDK.Runtime/ProcessingSourceExecutor.cs:134]: >//_CDS_ Should we "delayload" metadata tables? [](http://example.com/codeflow?start=12&length=46)

Yes :smile:

File: src/Microsoft.Performance.Toolkit.Engine/Engine.cs [hunk 2]
```diff
@@ -531,7 +530,7 @@ public void EnableTable(TableDescriptor descriptor)
             ITableExtensionReference tableReference = null;
             IEnumerable<ProcessingSourceExecutor> executorsWithTable = null;
 
-            if (!descriptor.IsInternalTable && this.Extensions.TablesById.TryGetValue(descriptor.Guid, out tableReference))
+            if (this.Extensions.TablesById.TryGetValue(descriptor.Guid, out tableReference))
```
Comment (trgibeau) [inline src/Microsoft.Performance.Toolkit.Engine/Engine.cs:533]: >(this.Extensions.TablesById.TryGetValue(descriptor.Guid, out tableReference)) [](http://example.com/codeflow?start=15&length=77)

should we double check that the descriptor has required cookers?

File: src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs [hunk 3]
```diff
@@ -582,21 +583,8 @@ public void BuildTable_InternalTables(EngineBuildTableTestCaseDto testCase)
...
-            Assert.AreEqual(0, builtTable2.TableCommands.Count());
-            Assert.IsNull(builtTable2.DefaultConfiguration);
-            Assert.IsNull(builtTable2.TableRowDetailsGenerator);
-
-            for (int i = 0; i < builtTable2.RowCount; i++)
-            {
-                Assert.AreEqual(i + 1, builtTable2.Columns.ElementAt(0).Project(i));
-                Assert.AreEqual($"Source5Row{i + 1}", builtTable2.Columns.ElementAt(1).Project(i));
-            }
+            // Metadata tables are no longer enabled by default and this should fail as it was not explicitly enabled
```
Comment (trgibeau) [inline src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs:586]: >// Metadata tables are no longer enabled by default and this should fail as it was not explicitly enabled [](http://example.com/codeflow?start=12&length=105)

Do we have an explicit enable test for metadata table?

**Comment by trgibeau** [comment]
>             bool allowInternalTables,

remove not needed

---
Refers to: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs:172 in 4387b48. [](commit_id = 4387b487f2107d61be3eea38254a1a51cc86887e, deletion_comment = False)

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs [hunk 4]
```diff
@@ -203,14 +183,6 @@ public bool IsInternalTable
...
                     out var tableIsDataAvailableFunc))
             {
-                if (tableDescriptor.IsInternalTable)
-                {
-                    if (!allowInternalTables)
-                    {
-                        return false;
-                    }
-                }
-
```
Comment (trgibeau) [inline src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs:213]: I upgraded WPA and XPerf with your changes, found that processor specified tables were being added as external. 

Updating to check if the build table action is null should work

File: src/Microsoft.Performance.SDK/Processing/CustomDataProcessor.cs [hunk 5]
```diff
@@ -265,7 +246,7 @@ protected virtual void EnableMetadataTables(IEnumerable<TableDescriptor> metadat
         /// </param>
         protected abstract void BuildTableCore(
             TableDescriptor tableDescriptor,
-            Action<ITableBuilder, IDataExtensionRetrieval> createTable,
+            Action<ITableBuilder> buildTable,
```
Comment (mslukebo) [inline src/Microsoft.Performance.SDK/Processing/CustomDataProcessor.cs:249]: can we fix up the warnings in this file

File: src/Microsoft.Performance.Toolkit.Engine/Engine.cs (hunk 2 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine/Engine.cs:533]: These are pulled from the data extension repo, so they should be extensions. But I can add a debug check

File: src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs (hunk 3 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs:586]: added

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs (hunk 4 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs:213]: This should be fixed now. Please let me know if there are still issues.

**Comment by mslukebo** [comment]
Can we remove `CustomDataProcessor.EnableMetadataTables`?

**Review by trgibeau — DISMISSED** [review]
:shipit:

File: src/Microsoft.Performance.SDK/Processing/IProcessorEnvironment.cs [hunk 6]
```diff
@@ -44,20 +43,5 @@ public interface IProcessorEnvironment
...
-        ///     extension tables, such as a metadata table that requires a source data cooker.
-        /// </summary>
-        /// <param name="processor">
-        ///     Custom data processor that supports data extensions. This should be the data processor to which this 
-        ///     <see cref="IProcessorEnvironment"/> was passed.
-        /// </param>
-        /// <returns>
-        ///     An instance of <see cref="IDataProcessorExtensibilitySupport"/>.
-        /// </returns>
-        IDataProcessorExtensibilitySupport CreateDataProcessorExtensibilitySupport(ICustomDataProcessorWithSourceParser processor);
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/IProcessorEnvironment.cs:61]: This was to help data cookers handle internal tables. It prepared the IDataExtensionRetrieval to enable them to build the table.

File: src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs [hunk 7]
```diff
@@ -42,33 +41,32 @@ public TableExtensionReference(TableExtensionReference other)
...
             {
                 this.AddRequiredDataCooker(dataCookerPath);
             }
 
-            foreach (var dataProcessorId in tableDescriptor.RequiredDataProcessors)
-            {
-                this.AddRequiredDataProcessor(dataProcessorId);
-            }
+            // TODO: __SDK_DP__
+            //foreach (var dataProcessorId in tableDescriptor.RequiredDataProcessors)
```
[the remark this answers is not in the document]
Comment (zachnew-msft) [reply src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Tables/TableExtensionReference.cs:66]: This was probably missed in the previous PR that disabled Data Processors.

File: src/Microsoft.Performance.SDK/Processing/ProcessingSource.cs [hunk 8]
```diff
@@ -320,7 +319,9 @@ private void InitializeAllTables(ITableConfigurationsSerializer tableConfigSeria
                     this.metadataTables.Add(table.Descriptor);
                 }
 
-                this.allTables[table.Descriptor] = table.BuildTable ?? GetTableBuildAction(table.Descriptor.Type);
+                this.allTables[table.Descriptor] = table.BuildTable != null
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/ProcessingSource.cs]: I don't think this is applicable anymore as allTables is now just a hashSet. please lemme know if i'm misreading this in github.

File: src/Microsoft.Performance.SDK/Processing/ProcessingSource.cs [hunk 9]
```diff
@@ -290,7 +289,7 @@ protected virtual void SetLoggerCore(ILogger logger)
         /// </summary>
         /// <param name="type">Type associated with the table</param>
         /// <returns>An action to build the table, or null</returns>
-        protected virtual Action<ITableBuilder, IDataExtensionRetrieval> GetTableBuildAction(
+        protected virtual Action<ITableBuilder> GetTableBuildAction(
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/ProcessingSource.cs]: I think this was resolved.

File: src/Microsoft.Performance.SDK/Processing/TableDescriptor.cs [hunk 10]
```diff
@@ -29,7 +29,14 @@ public sealed class TableDescriptor
         public const TableLayoutStyle DefaultLayoutStyle = TableLayoutStyle.GraphAndTable;
 
         private readonly HashSet<DataCookerPath> dataCookers;
-        private readonly HashSet<DataProcessorId> dataProcessors;
+
+        // TODO: __SDK_DP__
+        //private readonly HashSet<DataProcessorId> dataProcessors;
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/TableDescriptor.cs:34]: yeah, for issue #122 we decided to just comment these out for now

File: src/Microsoft.Performance.SDK.Tests/TableDescriptorUtils.cs [hunk 11]
```diff
@@ -15,7 +15,7 @@ internal static class TableDescriptorUtils
             ITableConfigurationsSerializer serializer,
             params Type[] types)
         {
-            CreateTableDescriptors(serializer, out var descriptors, out _, out _, types);
+            CreateTableDescriptors(serializer, out var descriptors, out var _, types);
```
Comment (zachnew-msft) [inline src/Microsoft.Performance.SDK.Tests/TableDescriptorUtils.cs:18]: nit: out _ is the syntax for discarding an out parameter (if we are truly not using it)

https://docs.microsoft.com/en-US/dotnet/csharp/fundamentals/functional/discards

File: src/Microsoft.Performance.SDK/Processing/TableDiscovery.cs [hunk 12]
```diff
@@ -6,16 +6,17 @@
...
 {
     /// <summary>
-    ///     Provides common <see cref="ITableProvider"/> implementations.
+    ///     Provides common <see cref="IProcessingSourceTableProvider"/> implementations.
     /// </summary>
     public static class TableDiscovery
     {
         /// <summary>
-        ///     Creates a new <see cref="ITableProvider"/> instance that will
+        ///     Creates a new <see cref="IProcessingSourceTableProvider"/> instance that will
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/TableDiscovery.cs:20]: oh, i missed some of these. I didn't realize there were so many. thanks.

File: src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine.Tests/ToolkitEngineTests.cs:586]: This has been removed, and there is a test for that.

File: src/Microsoft.Performance.SDK.Tests/TableDescriptorUtils.cs (hunk 11 above)
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK.Tests/TableDescriptorUtils.cs]: thx

**Review by zachnew-msft — DISMISSED** [review]

File: src/Microsoft.Performance.SDK/Processing/TableDiscovery.cs [hunk 13]
```diff
@@ -167,15 +174,18 @@ public static ITableProvider CreateForNamespace(string tableNamespace, Assembly[
                     out TableDescriptor td,
                     out Action<ITableBuilder, Extensibility.IDataExtensionRetrieval> buildTable))
                 {
-                    s.Add(new DiscoveredTable(td, buildTable));
+                    if (!td.RequiresDataExtensions() && buildTable == null)
```
[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.SDK/Processing/TableDiscovery.cs:192]: && is correct

**Review by trgibeau — DISMISSED** [review]
:shipit:

File: src/Microsoft.Performance.Toolkit.Engine.Tests/TestCookers/Source123/Source123Processor.cs [hunk 14]
```diff
@@ -13,27 +10,25 @@ public sealed class Source123Processor
...
+            if (tableDescriptor.Equals(Source123Table.TableDescriptor))
             {
                 var table = new Source123Table();
                 table.Build(tableBuilder);
             }
             else
             {
-                base.BuildTableCore(tableDescriptor, createTable, tableBuilder);
-            }            
+                base.BuildTableCore(tableDescriptor, tableBuilder);
```
[the remark this answers is not in the document]
Comment (mslukebo) [reply src/Microsoft.Performance.Toolkit.Engine.Tests/TestCookers/Source123/Source123Processor.cs:30]: for context, if I'm understanding the changes correctly we _don't_ want custom data processors to ever be building extended tables. What was happening in this test processor before was it was manually building Source123Table (a non-extension table), and then delegating everything else (extension tables) off to the base class. The base class _was_ just calling the build table action.

With our changes, we expect to never get extended tables in this method call.

[the remark this answers is not in the document]
Comment (jmaxson-ms) [reply src/Microsoft.Performance.Toolkit.Engine.Tests/TestCookers/Source123/Source123Processor.cs]: thanks

**Review by trgibeau — APPROVED** [review]

**Review by mslukebo — APPROVED** [review]

**Pull request merged**